### PR TITLE
[MIG 16.0 product_barcode_constraint_per_company

### PR DESCRIPTION
**Pull request by PierrickBrun**
some rework as this constraint was modified in odoo: https://github.com/odoo/odoo/commit/9584f80a48e46d300a7699391e610611c15377cf

### CODE REVIEW

File: product_barcode_constraint_per_company/readme/ROADMAP.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Due to some limitation of the framework, this module has to add a new stored field
+``tmpl_company_id`` on the ``product.product`` model related to the ``company_id``
+of the related ``product.template``.
```
Comment (legalsylvain) [inline product_barcode_constraint_per_company/readme/ROADMAP.rst]: This one should be removed.

File: product_barcode_constraint_per_company/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+# @author: Sylvain LE GAL (https://twitter.com/legalsylvain)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+{
+    "name": "Product Barcode Constraint per Company",
+    "version": "16.0.1.0.1",
+    "category": "Product",
+    "summary": "Change the product barcode constraint"
+    ", allowing the same barcode for differents companies",
+    "author": "GRAP, Odoo Community Association (OCA)",
```
Comment (legalsylvain) [inline product_barcode_constraint_per_company/__manifest__.py:11]:
```suggestion
    "author": "GRAP, Odoo Community Association (OCA)",
    "maintainers": ["legalsylvain"],
```

**Review by legalsylvain — APPROVED** [review]
minor comment. Otherwise, LGTM. 
thanks for porting this module !

**Comment by PierrickBrun** [comment]
I've added a glue module for product_multi_barcode

File: product_multi_barcode_constraint_per_company/models/product_barcode.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+from odoo import models
+
+
+class ProductBarcode(models.Model):
+    _inherit = "product.barcode"
+
+    def _get_duplicates(self, barcodes_to_check):
+        self.ensure_one()
+        barcodes_to_check = super()._get_duplicates(barcodes_to_check)
+        res = self.browse()
```
Comment (legalsylvain) [inline product_multi_barcode_constraint_per_company/models/product_barcode.py]: Why this line. We are in a self.ensure_one context.

Comment (PierrickBrun) [reply product_multi_barcode_constraint_per_company/models/product_barcode.py]: I don't want `self` to be included in the result

File: product_multi_barcode_constraint_per_company/models/product_barcode.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        for barcode in barcodes_to_check:
+            if (
+                not barcode.company_id
+                or not self.company_id
+                or barcode.company_id == self.company_id
+            ):
+                res |= barcode
```
Comment (rousseldenis) [inline product_multi_barcode_constraint_per_company/models/product_barcode.py]: @PierrickBrun for performances reasons:

```suggestion
            barcode_ids = set()
            for (...):
                barcode_ids.add(barcode.id)
            return self.browse(barcode_ids)
```

Comment (PierrickBrun) [reply product_multi_barcode_constraint_per_company/models/product_barcode.py]: Change done, this is the recordset modification that is bad performance-wise ?

**Comment by bealdav** [comment]
cc @rousseldenis maybe just have a look here

File: product_multi_barcode_constraint_per_company/models/product_barcode.py (hunk 4 above)
Comment (rousseldenis) [reply product_multi_barcode_constraint_per_company/models/product_barcode.py]: Yes, union is very slow.

Especially due to models name comparisons I think https://github.com/odoo/odoo/blob/16.0/odoo/models.py#L5790

**Review by rousseldenis — APPROVED** [review]
Code review

**Review by nayatec — APPROVED** [review]

**Comment by nayatec** [comment]
Hello mainteners, is there anything left to do for this PR to be merged ?

**Comment by bealdav** [comment]
I've just done a rebase.

If all is green I guess we can merge ? cc @rousseldenis

**Pull request merged**